src/dune_engine/load_rules.ml

```diff
@@ -660,6 +661,11 @@ end = struct
       Memo.return
         (Loaded.Build_under_directory_target { directory_target_ancestor })
     | Normal { rules; build_dir_only_sub_dirs; directory_targets } ->
+      let build_dir_only_sub_dirs =
+        Path.Local.Map.find build_dir_only_sub_dirs
+          (Path.Source.to_local sub_dir)
+        |> Option.value ~default:Subdir_set.empty
```

Comment:
I don't get how that would work if you're combining multiple `Build_config.Rules.t` where both come from different directories but have `Path.Local.root` keys.

Reply:
Roughly like this:

```ocaml
let combine (dir, map) (dir', map') =
  match Path.Build.descendant ~of_:dir dir' with
  | None -> (* impossible *)
  | Some p ->
    Path.Local.Map.foldi map' ~init:map ~f:(fun path dirs acc ->
      Path.Local.Map.add_exn acc (Path.Build.append_local p path) dirs)
```

Reply:
As I see it now, the choice comes to whether we want to pass `~dir` when 1. combining or when 2. creating the `Build_config.Rules.t` record.

Reply:
which one is that? I'd argue that my solution is simpler because it uses the directory we already have as the key in the map, rather than piecing it together afterwards (and requiring the user of the API to pass the extra directory arg)

Reply:
> which one is that? I'd argue that my solution is simpler because it uses the directory we already have as the key in the map

The problem is that this allows invalid keys in the map. So it would need an extra validation step.

> and requiring the user of the API to pass the extra directory arg)

Why should the user of the API pass the extra directory arg? When we evaluate the user callback we know exactly what the directory is.

Reply:
I think I understand. What paths are wrong in my case? The subdirectories that don't end up containing rules?

Reply:
In your case? You mean in how you use the API? I don't think there's anything wrong, it seems like you're using it correctly. What I mean is that it's possible to use this API is correctly, for example by passing `Path.Local.root` as the key from any other directory. While with my suggestion, `Path.Local.root` corresponds to the current directory and hence it's only possible to provide directories under it, as expected.

Reply:
I changed this PR to your suggestion, and I feel like it's a lot easier to follow what's happening when `Gen_rules` needs only to worry about looking up sub-directories for `Path.Local.root`.

I pushed it as a separate commit so that we can still have a chance to compare the 2 alternatives ([this commit](https://github.com/ocaml/dune/pull/7207/commits/a21a577428e292cf96a6be32df875c4a705dd2c4) vs [entire diff](https://github.com/ocaml/dune/pull/7207/files))